### ISSUE TITLE
[MPS] rearrange() returns a tensor mostly filled with NaN, until you look at it twice - regression in nightly

### ISSUE THREAD

**Issue by Birch-san** [pytorch/pytorch#85516]
### 🐛 Describe the bug

impacts [stable-diffusion's CrossAttention](https://github.com/Birch-san/stable-diffusion/blob/a49bf99c4ab4f25c157b7a92079a11908d9a2d1f/ldm/modules/attention.py#L180) during training — shown in debugger [here](https://github.com/invoke-ai/InvokeAI/issues/517#issuecomment-1255618027).  
on pytorch nightly 1.13.0.dev20220917. doesn't happen on stable 1.12.1.

```python
from torch import ones, float32
from einops import rearrange
from torch.nn import Linear

to_q = Linear(768, 320, bias=False, device='mps')
to_k = Linear(768, 320, bias=False, device='mps')
to_v = Linear(768, 320, bias=False, device='mps')

context = ones([1, 77, 768], device='mps', dtype=float32)
q = to_q(context)
k = to_k(context)
v = to_v(context)
q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> (b h) n d', h=8), (q, k, v))
print(q.isnan().sum())
# tensor(21440, device='mps:0')
print(q.isnan().sum())
# tensor(0, device='mps:0')
```

can [be mitigated](https://github.com/Birch-san/stable-diffusion/commit/5f470010eafb7069fd7ab034c176e2c4d91f3604) by detaching the tensor before interacting with it:

```python
print(q.detach().isnan().sum())
```

is it related at all to this "fails first time, works second time" bug?  
https://github.com/pytorch/pytorch/issues/85224

### Versions

```
PyTorch version: 1.13.0.dev20220917
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] pytorch-lightning==1.4.2
[pip3] torch==1.13.0.dev20220917
[pip3] torch-fidelity==0.3.0
[pip3] torchdiffeq==0.2.3
[pip3] torchmetrics==0.6.0
[pip3] torchtyping==0.1.4
[pip3] torchvision==0.14.0.dev20220917
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch-lightning         1.4.2                    pypi_0    pypi
[conda] torch                     1.13.0.dev20220917          pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchtyping               0.1.4                    pypi_0    pypi
[conda] torchvision               0.14.0.dev20220917          pypi_0    pypi
```

cc @kulinseth @albanD

**Comment by Birch-san** [comment]
`.detach()` may not be "special". the following also work as workarounds:

```python
print(q.contiguous().isnan().sum())
```

```python
print((q+0).isnan().sum())
```

**Comment by malfet** [comment]
I suspect its similar to einsum issue, as they share a lot of codebase.

**Comment by DenisVieriu97** [comment]
Fixed by Nikita's (@malfet) change from https://github.com/pytorch/pytorch/pull/85689 - latest pytorch nightly (1.13.0.dev20220929) already contains the fix. @Birch-san, could you please give a try and let me know if you see any other issues? Thanks!
